elytron/WFCORE-4360-CredentialStore_Expression_Resolution.adoc

```diff
@@ -0,0 +1,248 @@
+ 
+Beyond this the primary integration will be performed within the WildFly Core project as this enhancement is predominantly in relation to management model expression resolution.
+
+Additionally all documentation lives within the WildFly project so updates will be applied there accordingly.
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+The general intent for this enhancement is to develop some form of configurable expression resolver.
+
+The expression resolver will reference a specific alias within a specific credential store, this credential will be used to decrypt the expression which was previously encrypted.  The expression resolver will also contain configuration such as the algorithm to use.
+
+At all stages where installed `java.security.Provider` instances are used to obtain implementation it must remain possible to make use of alternative / custom provider implementations without being restricted exclusively to the implementations provided with the Java environment.
+
+==== Credential Store
+
+This enhancement will exclusively make use of `AES` `javax.crypt.SecretKey` instances, these do not require the use of a `javax.crypto.SecretKeyFactory` and can be generated using random key material directly using the `javax.crypto.spec.SecretKeySpec` class.  The generated `SecretKey` will be stored in a `CredentialStore`.
+
+It will be possible to generate, export, and import the SecretKey instances using both the command line tool and management operations.
+
+The Elytron tool `credential-store` command already has an `--add` action, rather than overloading this action for different types of key and different operations which will become complex very quickly a new set of actions will be added: -
+
+ * `--generate-secret-key`
+ * `--export-secret-key`
+ * `--import-secret-key`  
+
+The `--generate-secret-key` action will take the alias to store the key as an argument in the same pattern as `add` it will also take the following additional option: -
+
+ * `--size` - The key size (number of bits).
+ 
+The `--export-secret-key` action will take the alias of the previously stored key as an argument, no further option will be required.
+
+The `--export-secret-key` action will return a single Base64 value which is the key in it's encoded form.
+
+The `--import-secret-key` action will take the alias to store the key as an argument, it will also take the following additional option: -
+
+ * `--key` - The key to import encoded using Base64.
+
+If the `--key` option is not specified the user will be prompted to enter the key interactively, this will avoid the key being seen in the current running processes and avoid the key being cached in the users command line history.
+
+NOTE: For `generate-secret-key` and `import-secret-key` if an entry already exists for that alias / algorithm combination the behaviour will be the same as the `add` action in relation to replacing the entry or reporting an error.  (To be checked)
```

Comment:
@darranl Should this note still be here if we are going to support just the AES?

Reply:
Yes, I think this can remain - the credential store already supports the different permutations so that is not new behaviour being added by this RFE.